sympy/polys/matrices/linsolve.py

```diff
@@ -158,52 +159,122 @@ def _linear_eq_to_dict(eqs, syms):
+    is True, any nonlinearity no appearing already in ``s`` will raise a
+    PolyNonlinearError. When ``strict`` is False, only nonlinearity that
+    appears as a product of items from ``s`` in the expression will raise
+    an error (whether the expression is expanded or not).
+
+    Examples
+    ========
+
+    >>> from sympy.polys.matrices.linsolve import _le2d
+    >>> from sympy.abc import x, y, z
+    >>> strict = True
+    >>> _le2d(2*x + 3*y + z, {x, y}, strict)
+    {1: z, x: 2, y: 3}
+
+    The following does not raise an error because ``x**2`` does not appear in
+    ``x``:
+
+    >>> _le2d(x*(x**2 + 1), {x**2}, strict)
+    {1: x, x**2: x}
+
+    But the reverse would raise an error because ``x`` is in ``x**2``; but the
+    error can be supporessed, if desired, by using ``strict=False`. This might
+    be most useful when wanting to allow ``f(x)`` and ``f(x).diff()`` to be
+    treated as independent of each other.
+
+    >>> _le2d(x*(x**2 + 1), {x}, not strict)
+    {x: x**2 + 1}
+    """
+    assert isinstance(e, Basic)
+    assert type(s) is set
+    assert strict in (True, False)
+    if e in s:
+        return {e: S.One}
+    if e.is_Mul:
+        c = []
+        d = None
+        for ei in e.args:
+            di = _le2d(ei, s, strict)
+            if len(di) == 1 and 1 in di:
+                c.append(ei)
+            elif d:
+                raise PolyNonlinearError('nonlinear term: %s' % e)
             else:
-                raise PolyNonlinearError
-        coeff = Mul(*coeff_list)
-        if terms is None:
-            return coeff, {}
-        else:
-            terms = {sym: coeff * c for sym, c in terms.items()}
-            return  coeff * terms_coeff, terms
-    elif a.is_Equality:
-        return _lin_eq2dict(a.lhs - a.rhs, symset)
-    elif not a.has_free(*symset):
-        return a, {}
-    else:
-        raise PolyNonlinearError
+                d = di
+        c = Mul._from_args(c)
+        return {k: c*v for k, v in d.items()} if d else {1: c}
+    if e.is_Add:
+        c, t = [], []
+        for ei in e.args:
+            if ei in s:
+                t.append(({ei: 1}, 0))
+            elif not ei.has_free_arg(s):
+                c.append(ei)
+            elif not ei.is_Mul:
+                if strict:
+                    raise PolyNonlinearError('nonlinear term: %s' % ei)
+                c.append(ei)
+            else:
+                d = _le2d(ei, s, strict)
+                if len(d) == 1 and 1 in d:
+                    c.append(ei)
+                else:
+                    t.append((d, d.pop(1, 0)))
+        rv = {1: [Add._from_args(c)]}
+        for d, ci in t:
+            for k, v in d.items():
+                rv.setdefault(k, []).append(v)
+            rv[1].append(ci)
+        rv = {k: Add(*v) for k, v in rv.items()}
+        return rv
+    if not e.has_free_arg(s):
+        return {1: e}
+    if not strict:
+        return {1: e}
```

Comment:
Apart from these lines is it necessary to change anything else about this function?

I find the changed code harder to read than the original. There are more branch points and I don't really understand why. The names of variables and the function name itself are also cryptic.

Reply:
Yes, there is more to making non-strict work. For example, those two lines aren't enough to make `x*(1/x + 1)` be recognized as `0, {x: 1 + 1/x}` or `(F(x)*F(x).diff(x)` to be recognized as `{f(x): Derivative(f(x), x)}`.

Reply:
I tried the following diff relative to master:
```diff
diff --git a/sympy/polys/matrices/linsolve.py b/sympy/polys/matrices/linsolve.py
index 75ae26d..27192a7 100644
--- a/sympy/polys/matrices/linsolve.py
+++ b/sympy/polys/matrices/linsolve.py
@@ -168,7 +168,7 @@ def _linear_eq_to_dict_inner(eqs, syms):
     return eqsdict, eqs_rhs
 
 
-def _lin_eq2dict(a, symset):
+def _lin_eq2dict(a, symset, strict=False):
     """Efficiently convert a linear equation to a dict of coefficients"""
     if a in symset:
         return S.Zero, {a: S.One}
@@ -205,5 +205,7 @@ def _lin_eq2dict(a, symset):
         return _lin_eq2dict(a.lhs - a.rhs, symset)
     elif not a.has_free(*symset):
         return a, {}
+    elif not strict:
+        return a, {}
     else:
         raise PolyNonlinearError
```
With that:
```python
In [1]: from sympy.polys.matrices.linsolve import _linear_eq_to_dict

In [2]: _linear_eq_to_dict([f(x)*f(x).diff(x)], [f(x)])
Out[2]: 
⎛⎡⎧      d       ⎫⎤     ⎞
⎜⎢⎨f(x): ──(f(x))⎬⎥, [0]⎟
⎝⎣⎩      dx      ⎭⎦     ⎠

In [3]: _linear_eq_to_dict([f(x)*f(x).diff(x)], [f(x).diff(x)])
Out[3]: 
⎛⎡⎧d             ⎫⎤     ⎞
⎜⎢⎨──(f(x)): f(x)⎬⎥, [0]⎟
⎝⎣⎩dx            ⎭⎦     ⎠

In [4]: _linear_eq_to_dict([x*(1 + 1/x)], [x])
Out[4]: 
⎛⎡⎧       1⎫⎤     ⎞
⎜⎢⎨x: 1 + ─⎬⎥, [0]⎟
⎝⎣⎩       x⎭⎦     ⎠
```

Reply:
I had misapplied the lines that I thought you were referring to. Let's give it a go. Local tests show that it is faster. I also will replace the `has_free` with `has_free_arg`.